### PR TITLE
Add and use serialization and dictionary extensions

### PR DESCRIPTION
**Pull request by mattjohnsonpint**
Following up feedback on #1834 

- Uses Apple's `NSJsonSerialization` for JSON serialization of `NSObject` and derived types.
- Consolidates `NSDictionary` conversions into extension methods.

#skip-changelog

### CODE REVIEW

File: src/Sentry/Platforms/iOS/Extensions/BreadcrumbExtensions.cs [hunk 1]
```diff
@@ -1,43 +1,26 @@
...
-        var items = breadcrumb.Data as IEnumerable<KeyValuePair<NSString, NSObject?>>;
-        var data = items?.ToDictionary(
-            x => (string)x.Key,
-            x => x.Value?.ToString() ?? "");
-
-        return new Breadcrumb(
+    public static Breadcrumb ToBreadcrumb(this SentryCocoa.SentryBreadcrumb breadcrumb, IDiagnosticLogger? logger) =>
+        new(
             breadcrumb.Timestamp?.ToDateTimeOffset(),
             breadcrumb.Message,
```
Comment (SimonCropp) [inline src/Sentry/Platforms/iOS/Extensions/BreadcrumbExtensions.cs:10]: not that u changed this. but i get null warnings for message and type 
image

should we default them to something?

File: src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs [hunk 2]
```diff
@@ -1,8 +1,110 @@
...
+    {
+        if (dict == null)
+        {
+            return new Dictionary<string, string>();
+        }
+
+        var result = new Dictionary<string, string>(capacity: (int)dict.Count);
+        foreach (var key in dict.Keys)
+        {
+            var value = dict[key];
```
Comment (SimonCropp) [inline src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs:60]:
```suggestion
        foreach (var keyValue in dict)
        {
            var value = keyValue.Value;
            var key = keyValue.Key;
```

no need to do a dictionary lookup

File: src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs [hunk 3]
```diff
@@ -1,8 +1,110 @@
...
+                d.Values.ToArray(),
+                d.Keys.ToArray());
+    }
+
+    public static NSDictionary<NSString, NSObject>? ToNullableNSDictionary<TValue>(
+        this IEnumerable<KeyValuePair<string, TValue>> dict)
+    {
+        var d = dict.ToNSDictionary();
+        return d.Count == 0 ? null : d;
+    }
```
Comment (SimonCropp) [inline src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs:109]:
```suggestion
    public static NSDictionary<NSString, NSObject>? ToNullableNSDictionary<TValue>(
        this ICollection<KeyValuePair<string, TValue>> dict)
    {
        if (dict.Count == 0)
        {
            return null;
        }

        return dict.ToNSDictionary();
    }
```

File: src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs [hunk 4]
```diff
@@ -1,8 +1,110 @@
...
+        IDiagnosticLogger? logger = null)
+        where TValue : NSObject
+    {
+        if (dict == null)
+        {
+            return new Dictionary<string, string>();
+        }
+
+        var result = new Dictionary<string, string>(capacity: (int)dict.Count);
+        foreach (var key in dict.Keys)
```
Comment (SimonCropp) [inline src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs:58]: u cant enumerate key-value pairs in `NSDictionary`?

File: src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs [hunk 5]
```diff
@@ -1,8 +1,110 @@
...
+        return result;
+    }
+
+    public static Dictionary<string, string>? ToNullableStringDictionary<TValue>(
+        this NSDictionary<NSString, TValue>? dict,
+        IDiagnosticLogger? logger = null)
+        where TValue : NSObject
+    {
+        var d = dict?.ToStringDictionary(logger);
+        return d?.Count is null or 0 ? null : d;
```
Comment (SimonCropp) [inline src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs:82]:
```suggestion
        if (dict is null || dict.Length == 0)
        {
            return null;
        }

        return dict.ToStringDictionary(logger);
```

File: src/Sentry/Platforms/iOS/Extensions/BreadcrumbExtensions.cs (hunk 1 above)
Comment (mattjohnsonpint) [reply src/Sentry/Platforms/iOS/Extensions/BreadcrumbExtensions.cs:10]: They are nullable.  Intellisense failure.

File: src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs (hunk 4 above)
Comment (mattjohnsonpint) [reply src/Sentry/Platforms/iOS/Extensions/CocoaExtensions.cs:58]: Not easily.

**Review by SimonCropp — APPROVED** [review]

**Pull request merged**